hoot-core/src/main/cpp/hoot/core/io/OsmApiDbAwareHootApiDbWriter.cpp

```diff
@@ -0,0 +1,290 @@
+  {
+    // This is only called when adding nodes for a way.  If a way has a node we
+    //    did not successfully create a mapping for when importing nodes,
+    //    we can't continue
+    if (_nodeRemap.count(nids[i]) == 1)
+    {
+      result[i] = _nodeRemap.at(nids[i]);
+    }
+    else
+    {
+      //TODO: This isn't good, but haven't figured out how to get around it yet.
+      if (ConfigOptions().getHootapiDbWriterOsmapidbUrl().isEmpty())
+      {
+        throw HootException(QString("Requested ID remap for node " +  QString::number(nids[i])
+          + QString(" but it did not exist in mapping table")));
+      }
+      else
+      {
+        result[i] = nids[i];
+      }
+    }
+  }
+
+  return result;
+}
+
+void OsmApiDbAwareHootApiDbWriter::writePartial(const shared_ptr<const Node>& n)
+{
+  bool countChange = true;
+
+  LOG_VARD(n->getId());
+
+  Tags t = n->getTags();
+  _addElementTags(n, t);
+
+  if (n->getId() < 1)
+  {
+    bool alreadyThere = _nodeRemap.count(n->getId()) != 0;
+    LOG_VARD(alreadyThere);
+    long nodeId = _getRemappedElementId(n->getElementId());
+    LOG_VARD(nodeId);
+    if (alreadyThere)
+    {
+      _hootdb.updateNode(nodeId, n->getY(), n->getX(), n->getVersion() + 1, t);
+    }
+    else
+    {
+      _hootdb.insertNode(nodeId, n->getY(), n->getX(), t);
+    }
+  }
+  else
+  {
+    _hootdb.insertNode(n->getId(), n->getY(), n->getX(), t);
+  }
+
+  if (countChange)
+  {
+    _countChange();
+    _nodesWritten++;
+  }
+}
+
+void OsmApiDbAwareHootApiDbWriter::writePartial(const shared_ptr<const Way>& w)
+{
+  long wayId;
+
+  Tags tags = w->getTags();
+  _addElementTags(w, tags);
+
+  if (w->getId() < 1)
+  {
+    bool alreadyThere = _wayRemap.count(w->getId()) != 0;
+    LOG_VARD(alreadyThere);
+    wayId = _getRemappedElementId(w->getElementId());
+    LOG_VARD(wayId);
+    if (alreadyThere)
+    {
+      _hootdb.updateWay(wayId, w->getVersion() + 1, tags);
+    }
+    else
+    {
+      _hootdb.insertWay(wayId, tags);
+    }
+  }
+  else
+  {
+    wayId = w->getId();
+    _hootdb.insertWay(w->getId(), tags);
+  }
+
+  if (_remapIds == true)
+  {
+    _hootdb.insertWayNodes(wayId, _remapNodes(w->getNodeIds()));
+  }
+  else
+  {
+    _hootdb.insertWayNodes(wayId, w->getNodeIds());
+  }
+
+  _countChange();
+
+  _waysWritten++;
+}
+
+void OsmApiDbAwareHootApiDbWriter::writePartial(const shared_ptr<const Relation>& r)
+{
+  long relationId;
+
+  Tags tags = r->getTags();
+  _addElementTags(r, tags);
+
+  if (!r->getType().isEmpty())
+  {
+    tags["type"] = r->getType();
+  }
+
+  if (r->getId() < 1)
+  {
+    relationId = _getRemappedElementId(r->getElementId());
+    LOG_VARD(relationId);
+
+    //LOG_DEBUG("Inserting relation with source ID = " <<
+              //QString::number(r->getId()) << " which maps to DB ID = " <<
+              //QString::number(relationId) );
+
+    _hootdb.insertRelation(relationId, tags);
+  }
+  else
+  {
+    _hootdb.insertRelation(r->getId(), tags);
```

Comment:
Could this clash if the DB already contains an element with this ID? Are you assuming that all positive IDs came from this DB? If so, shouldn't this be an update rather than an insert?